Etcd2Topo: Use node's ModRevision consistently for in-memory topo.Version value

mattlord opened the pull request:
## Description

Etcd has several version related fields/values in each key's header: `Revision`, `ModRevision`, and `Version`. These are related but distinct things:

  - https://github.com/etcd-io/etcd/issues/6518
  - https://etcd.io/docs/v3.5/learning/data_model/

To summarize:

  - `Revision`: a logical clock used to track changes to the etcd keyspace (a transaction / snapshot ID)
  - `Version`: a logical clock used to track the changes to an individual key _since it was last created_ (it is not monotonic)
  - `ModRevision`: the etcd keyspace revision number for the key at that point in time (the logical version of the key)

The `Revision` is what allows for consistent reads and writes of multiple keys (in a transaction). It allows you, for example, to do a consistent read / scan of N keys at that logical point-in-time.

In Vitess we also have awareness of each value and read/use these values in memory in the `etcd2topo` topo server implementation. There's also a generic and opaque `topo.Version` type — which can be set to either of the two lower level node/key specific values: `ModRevision` or `Version` — that we use for [strict serializability](https://etcd.io/docs/v3.5/learning/api_guarantees/).

In https://github.com/vitessio/vitess/pull/15701 I aligned the Vitess and `etcd2topo` lower level version related fields, variable names, and comments for watches: https://github.com/vitessio/vitess/pull/15701/files/f1fed6dd787feea26a4f4e13f660e8eaa91e64bb#diff-3f055ca1a4446966abba5a840e13c5c6a11e73121143a66b06070a356425d676

In later unrelated investigations, however, I realized that the use of `Version:  EtcdVersion(initial.Kvs[0].ModRevision)` in `Watch()` was intentional and consistent with how we're using `ModRevision` for the node's `topo.Version` value in [`file.go`](https://github.com/vitessio/vitess/blob/main/go/vt/topo/etcd2topo/file.go). The reason being that we use this to enforce [strict serializability](https://etcd.io/docs/v3.5/learning/api_guarantees/) when a `topo.Version` value is provided — failing the write if the `topo.Version` value provided is not equal to the current `ModRevision` of the key — and we return the new `topo.Version` value based on the new `ModRevision` value (technically we return the `Revision` value from the response header, but that would also be the `ModRevision` value for the key we updated in the transaction) on successful update: https://github.com/vitessio/vitess/blob/f27d287e0873d139ad2f71588068256e28a16f6a/go/vt/topo/etcd2topo/file.go#L48-L74 

We use the `ModRevision` value there as etcd's transaction response includes the new `ModRevision` from the commit 
(again, technically it's the `Revision` value from the response header, but that would also be the `ModRevision` value for the key we updated in the transaction) but NOT the new node/key `Version`. This makes sense because a transaction can update any arbitrary number of keys. So the only logical version that can be provided for that transaction is the keyspace `Revision`, which would be the `ModRevision` value _for each of the keys updated_ in the transaction. This is the logical point in time for the state of the data in etcd — equivalent to a GTID position/set in MySQL.

That is used e.g. with `Shard` records/keys and the `ShardInfo` in memory structure: https://github.com/vitessio/vitess/blob/f27d287e0873d139ad2f71588068256e28a16f6a/go/vt/topo/shard.go#L165-L222

One way for us to think about this: an etcd watch is equivalent to a (filtered) binlog stream with MySQL. We don’t know the version of a table or row, we only know their state at a logical point in time — the GTID position. The revision in etcd is that logical point in time / position in the stream of events/changes.

So with all of this in mind, we should also be using `ModRevision` for the `topo.Version` value in `Watch()` — and beyond that, we should be using it for all `topo.Version` values within the `etcd2topo` implementation. Please see the discussion [from here](https://github.com/vitessio/vitess/pull/15847#issuecomment-2096935831) on for further reasons why.

So in the end, this PR does just that: it uses the etcd `ModRevision` value for the logical `topo.Version` value of keys across the `etcd2topo` implementation and adds a test to check and enforce that for `Watch()`.

## Related Issue(s)

  - Follow-up to: https://github.com/vitessio/vitess/pull/15701

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

rohit-nayak-ps commented:
To clarify, the use of `Version` is not causing a functional issue in Vitess, right? This change is only to be consistent?

Since we are watching a specific key, it shouldn't matter if we use `Version` or `ModRevision` as both will increment anytime a key is updated.

deepthi commented:
> To clarify, the use of `Version` is not causing a functional issue in Vitess, right? This change is only to be consistent?
> 
> Since we are watching a specific key, it shouldn't matter if we use `Version` or `ModRevision` as both will increment anytime a key is updated.

Quote from the description:
this PR is really a 1 line revert from the noted PR.

So the way I'm reading this is that there was an unnecessary change in a previous PR which may actually cause incorrect behavior, and this PR is fixing that.

mattlord commented:
> To clarify, the use of `Version` is not causing a functional issue in Vitess, right? This change is only to be consistent?
> 
> Since we are watching a specific key, it shouldn't matter if we use `Version` or `ModRevision` as both will increment anytime a key is updated.

`Version` and `ModRevision` will both be increasing, yes. So you can use one or the other as long as you're consistent. They just aren't the same logical or literal value (as Revision is incremented each time the etcd keyspace is updated). So it matters which one you return, and the caller can then pass back in on subsequent topo calls, as we then check that this value is EQUAL TO the key's current ModRevision (if a `topo.Value` was passed in, e.g. to `Update()`).

AFAIK, the logical version within watches only comes in to play when we need to resume/restart the watch: https://github.com/vitessio/vitess/blob/7603abc53459065f4420c92a0ae29f5ccb1cad10/go/vt/topo/etcd2topo/watch.go#L79-L110

As you can see there, it's already using the revision value (etcd keyspace version) for both. BUT, in looking at that I noticed that in a relatively new usage of `watchData{}` from https://github.com/vitessio/vitess/pull/10906 it was using `Version` rather than `ModRevision`.  I can now understand WHY that was done at the time, but I've now updated that to use `ModRevision` just as the normal (non-recursive) `Watch` does for the same reasons as noted in the issue description.

rohit-nayak-ps commented:
OK, I just saw after my comment above that you have a new commit where everything is set as `ModRevision`. I guess this should be good. It will still be useful to have a unit test that shows that we get newer revisions in the watcher, since we think it is worthwhile to modify this behaviour.

mattlord commented:
> OK, I just saw after my comment above that you have a new commit where everything is set as `ModRevision`. I guess this should be good.

Yes, I noticed this too in answering your previous question and I updated things to be consistent (also updated the description for further clarification).

> It will still be useful to have a unit test that shows that we get newer revisions uniformly in the watcher, since we think it is worthwhile to modify this behaviour.

Sure, I can add one.

rohit-nayak-ps commented:
Thinking about this some more: the watchers are watching a specific key, right? So shouldn't we be passing `Version` which is key-specific, rather than the global `ModRevision`. For example, in a unit test, if the SrvVSchema is changed thrice, we can expect the watcher to stream us versions 1, 2 and 3, but we have no idea from the `ModRevision` whether we have received all the updates.

When will using `Version` everywhere be wrong?

mattlord commented:
> Thinking about this some more: the watchers are watching a specific key, right? So shouldn't we be passing `Version` which is key-specific, rather than the global `ModRevision`. For example, in a unit test, if the SrvVSchema is changed thrice, we can expect the watcher to stream us versions 1, 2 and 3, but we have no idea from the `ModRevision` whether we have received all the updates.

`ModRevision` *is* a metadata property of the key. That value is monotonically increasing. `Version` goes back to 0 if the key is deleted and recreated. The watchers have *never* guaranteed that you don't miss an intermediate value. They can't, as etcd will do compactions and a user can request it as well. This effectively prunes the history. So in various scenarios (including etcd failover) you will have to restart your watch and you will have missed intermediate values as they are no longer available and you start with the current/latest version.

https://github.com/vitessio/vitess/blob/96c43e40b86dd03b8a462507ebd82fbc5c927f8e/go/vt/topo/etcd2topo/watch.go#L63-L66

https://github.com/vitessio/vitess/blob/96c43e40b86dd03b8a462507ebd82fbc5c927f8e/go/vt/topo/etcd2topo/watch.go#L63-L66

And it's important to note that the etcd client library *only* takes revisions on `Get`,`Watch`,`Update` etc as there is no `clientv3.WithVersion`:
```
s.cli.Watch(watchCtx, nodePath, clientv3.WithRev(rev))
```

You're saying that you want to watch this key from this database snapshot. And this is the only way that you can coordinate across keys, whether you're using transactions, multiple watches, etc.

https://github.com/etcd-io/etcd/blob/35361955578322a55841c1ed962c53e28658d951/client/v3/watch.go#L53-L82

> When will using `Version` everywhere be wrong?

Wrong is not the correct word to use here IMO. We have to define the semantics and guarantees -- only then can we know the best option. I've made every attempt to explain why `ModRevision` *has* to be used in the places where we care most about the `topo.Version` value.  So we know that we can't use it *everywhere*. If you only mean everywhere *in the Watch* then I get the inclination. The downside I see there is that if you read the latest version of a key from the watcher, then update it, you won't be able to save that back in the topo if you provide the `topo.Version` value you have. So you'd be forced to do a blind write. 

So I take this in total to mean that the key's `Version` value is handy for humans but what really matters is its `ModRevision` value — as that's what is used for consistent reads and writes and thus what you pass when you want to read, watch, update a key.

mattlord commented:
After further investigation (see my previous comment: https://github.com/vitessio/vitess/pull/15847#issuecomment-2096935831) I think that we should use `Revision/ModRevision` everywhere as the `topo.Version` value as that seems to be the only thing that etcdv3 cares about. In this PR, we're already doing that with one exception:
```
❯ git grep -E 'Version|Revision' | grep etcd2topo | grep -v "Apache License, Version 2.0"
go/vt/topo/etcd2topo/election.go:		clientv3.WithSort(clientv3.SortByModRevision, clientv3.SortAscend),
go/vt/topo/etcd2topo/election.go:		clientv3.WithSort(clientv3.SortByModRevision, clientv3.SortAscend),
go/vt/topo/etcd2topo/election.go:	watcher := mp.s.cli.Watch(ctx, electionPath, clientv3.WithPrefix(), clientv3.WithRev(initial.Header.Revision))
go/vt/topo/etcd2topo/election.go:					clientv3.WithSort(clientv3.SortByModRevision, clientv3.SortAscend),
go/vt/topo/etcd2topo/file.go:func (s *Server) Create(ctx context.Context, filePath string, contents []byte) (topo.Version, error) {
go/vt/topo/etcd2topo/file.go:		If(clientv3.Compare(clientv3.Version(nodePath), "=", 0)).
go/vt/topo/etcd2topo/file.go:	return EtcdVersion(txnresp.Header.Revision), nil
go/vt/topo/etcd2topo/file.go:func (s *Server) Update(ctx context.Context, filePath string, contents []byte, version topo.Version) (topo.Version, error) {
go/vt/topo/etcd2topo/file.go:			If(clientv3.Compare(clientv3.ModRevision(nodePath), "=", int64(version.(EtcdVersion)))).
go/vt/topo/etcd2topo/file.go:			return nil, topo.NewError(topo.BadVersion, nodePath)
go/vt/topo/etcd2topo/file.go:		return EtcdVersion(txnresp.Header.Revision), nil
go/vt/topo/etcd2topo/file.go:	return EtcdVersion(resp.Header.Revision), nil
go/vt/topo/etcd2topo/file.go:func (s *Server) Get(ctx context.Context, filePath string) ([]byte, topo.Version, error) {
go/vt/topo/etcd2topo/file.go:	return resp.Kvs[0].Value, EtcdVersion(resp.Kvs[0].ModRevision), nil
go/vt/topo/etcd2topo/file.go:		results[n].Version = EtcdVersion(pairs[n].ModRevision)
go/vt/topo/etcd2topo/file.go:func (s *Server) Delete(ctx context.Context, filePath string, version topo.Version) error {
go/vt/topo/etcd2topo/file.go:			If(clientv3.Compare(clientv3.ModRevision(nodePath), "=", int64(version.(EtcdVersion)))).
go/vt/topo/etcd2topo/file.go:					return topo.NewError(topo.BadVersion, nodePath)
go/vt/topo/etcd2topo/lock.go:		If(clientv3.Compare(clientv3.Version(newKey), "=", 0)).
go/vt/topo/etcd2topo/lock.go:	return newKey, txnresp.Header.Revision, nil
go/vt/topo/etcd2topo/server.go:			MinVersion:         tls.VersionTLS12,
go/vt/topo/etcd2topo/version.go:// EtcdVersion is etcd's idea of a version.
go/vt/topo/etcd2topo/version.go:// It implements topo.Version.
go/vt/topo/etcd2topo/version.go:type EtcdVersion int64
go/vt/topo/etcd2topo/version.go:// String is part of the topo.Version interface.
go/vt/topo/etcd2topo/version.go:func (v EtcdVersion) String() string {
go/vt/topo/etcd2topo/watch.go:		// ModRevision is used for the topo.Version value as we get the new Revision value back
go/vt/topo/etcd2topo/watch.go:		Version: EtcdVersion(initial.Kvs[0].ModRevision),
go/vt/topo/etcd2topo/watch.go:	watcher := s.cli.Watch(watchCtx, nodePath, clientv3.WithRev(initial.Header.Revision))
go/vt/topo/etcd2topo/watch.go:		var rev = initial.Header.Revision
go/vt/topo/etcd2topo/watch.go:				rev = wresp.Header.GetRevision()
go/vt/topo/etcd2topo/watch.go:							Version:  EtcdVersion(ev.Kv.ModRevision),
go/vt/topo/etcd2topo/watch.go:		wd.Version = EtcdVersion(initial.Kvs[0].ModRevision)
go/vt/topo/etcd2topo/watch.go:	watcher := s.cli.Watch(watchCtx, nodePath, clientv3.WithRev(initial.Header.Revision), clientv3.WithPrefix())
go/vt/topo/etcd2topo/watch.go:		var rev = initial.Header.Revision
go/vt/topo/etcd2topo/watch.go:				rev = wresp.Header.GetRevision()
```

That one exception is here: https://github.com/vitessio/vitess/blob/96c43e40b86dd03b8a462507ebd82fbc5c927f8e/go/vt/topo/etcd2topo/file.go#L29-L38

The usage there makes sense to me because it's a special case when creating a key, we want to enforce that it doesn't exist (at lower layers it's a put whether we're creating a key or updating it).

So I'm going to add a test to confirm and enforce that we are using `ModRevision` in watches and that we observe the relevant semantics. This way we are at least consistent, testing the current behavior, and preventing (unintentional) changes going forward.

mattlord commented:
FWIW, here's an example: 
```
❯ etcdctl get -w json /testval
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":35,"raft_term":2}}

❯ etcdctl put /testval "foo"
OK

❯ etcdctl get -w json /testval
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":36,"raft_term":2},"kvs":[{"key":"L3Rlc3R2YWw=","create_revision":36,"mod_revision":36,"version":1,"value":"Zm9v"}],"count":1}

❯ etcdctl put /testval "foo2"
OK

❯ etcdctl get -w json /testval
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":37,"raft_term":2},"kvs":[{"key":"L3Rlc3R2YWw=","create_revision":36,"mod_revision":37,"version":2,"value":"Zm9vMg=="}],"count":1}

❯ etcdctl put /testval "foo3"
OK

❯ etcdctl get -w json /testval
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":38,"raft_term":2},"kvs":[{"key":"L3Rlc3R2YWw=","create_revision":36,"mod_revision":38,"version":3,"value":"Zm9vMw=="}],"count":1}

❯ etcdctl put -w json /testval "foo3"
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":39,"raft_term":2}}
❯ etcdctl put -w json /testval "foo3"
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":40,"raft_term":2}}
❯ etcdctl put -w json /testval "foo3"
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":41,"raft_term":2}}
❯ etcdctl put -w json /testval "foo3"
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":42,"raft_term":2}}
```

rohit-nayak-ps commented:
> Let's also look at an example since you seem to suspect that Version would give us better properties/semantics in determining order / sequence of operations and a key's relative logical version (is this key value newer than another one):

I do understand the differences between Version and ModRevision and that if a key is deleted we restart the versioning. 

My only context to ask about Version vs ModRevision is related to the watcher, since that is the code path this PR affects. My understanding in this context is:

If we don't care about the version of the watched object (has it been changed X times) and/or if it is a newly created key then it doesn't matter which one we use.

If it does matter, do we delete the watched key ever?  If not, it doesn't matter which one we use.

If so, does it matter to us that a key was deleted? If not,  it doesn't matter which one we use.

If so, we do want the `Version` (or both the `CreateRevision` and `ModRevision`) since it will tell us that we have a new key. 

So in terms of watching specific keys, I believe the `Version` gives us more information than `ModRevision`. Or possibly sending all the version numbers, `Version`, `CreateRevision` and `ModRevision` is most flexible.

Just to round off the discussion on this PR, from my side, I am fine with whatever the others decide we want to do now for the watch versioning since we don't seem to depend on the versions in this context anyway. 

We can always change it later if any specific functionality is needed that depends on this decision.

mattlord commented:
@rohit-nayak-ps and @deepthi thank you for the reviews and discussion! I've now thoroughly documented the reason for this PR and the reasoning for the changes made. I've also now added a unit test to confirm and enforce the new expected behavior.

I'm considering this work done unless I get any further review comments. Thanks again!
